add wsl setup docs and reorganize a new "dev" folder. (#678)
leanprover/lean4: doc/dev/testing.md

# Unit Testing

You can run the unit tests after completing a build using the following:

```bash
cd build/release/stage0

> ctest -j 4 --output-on-failure --timeout 300

Reviewer: There is some overlap here with `## Development Setup`. I like `make test` better because it ensures the build is up to date. Also, there should never be a need to run tests in stage 0. Not sure if we need the timeout specifier, it seems like some rare race condition in a single server test case on Windows.

Reviewer: Ah, maybe not that rare... we should fix the test, though I'm not exactly sure how given the circumstances. Perhaps we should activate excessive logging for it on CI and wait for it to fail again.

```

You can increase the parallelism from 4 to as many cores you have
available to decrease overall test time.  You can also increase the timeout
from 300 seconds if you need to.

Here is the summary of the test source code organization.

> All these tests are included by [~/src/shell/CMakeLists.txt](https://github.com/leanprover/lean4/blob/master/src/shell/CMakeLists.txt):

Reviewer: ```suggestion
All these tests are included by [src/shell/CMakeLists.txt](https://github.com/leanprover/lean4/blob/master/src/shell/CMakeLists.txt):

```

Author: fixed

- `tests/lean`: contains tests that come equipped with a
  .lean.expected.out file. The driver script `test_single.sh` runs
  each test and checks the actual output (*.produced.out) with the
  checked in expected output.

- `tests/lean/run`: contains tests that are run through the lean
  command line one file at a time. These tests only look for error
  codes and do not check the expected output even though output is
  produced, it is ignored.

> - `tests/lean/interactive`: are designed to test auto-completion

Reviewer: ```suggestion
- `tests/lean/interactive`: are designed to test server

```

Author: fixed

  requests at a given position in the input file. Each .lean files
  contain comments that indicate how to simulate a client request at
  that position. using a `--^` point to the line position. Example:

>     ```lean

Author: fixed

    open Foo in
    theorem tst2 (h : a ≤ b) : a + 2 ≤ b + 2 :=

>     Bla.
>         --^ textDocument/completion

Reviewer: ```suggestion
    Bla.
      --^ textDocument/completion

```

Author: fixed

    ```
    In this example, the test driver `test_single.sh` will simulate an
    auto-completion request at `Bla.`. The expected output is stored in
    a .lean.expected.out in the json format that is part of the
    [Language Server
    Protocol](https://microsoft.github.io/language-server-protocol/).

- `tests/lean/server`: Tests more of the lean `--server` protocol.
  There are just a few of them, and it uses .log files containing
  JSON.

- `tests/compiler`: contains tests that will run the lean compiler and
  build an executable that is executed and the output is compared to
  the .lean.expected.out file. This test also contains a subfolder
  `foreign` which shows how to extend Lean using C++.

- `tests/lean/trust0`: tests that run Lean in a mode that Lean doesn't
  even trust the .olean files (i.e., trust 0).

- `tests/bench`: contains performance tests.

> - `tests/plugin`: tests that one compiled lean program can call
>   another compiled lean program via the `--plugin` command line

Reviewer: ```suggestion
- `tests/plugin`: tests that compiled Lean code can be loaded
  into `lean` via the `--plugin` command line
```

Author: fixed

You can run the unit tests after completing a build using the following:

After [building lean](../make/index.md) you can run all the tests using
```
cd build/release
make test ARGS=-j4
```

Change the 4 to the maximum number of parallel tests you want to
allow. The best choice is the number of CPU cores on your machine as
the tests are mostly CPU bound.  You can find the number of processors
on linux using `cat /proc/cpuinfo | grep processor | wc -l` and on
Windows it is the NUMBER_OF_PROCESSORS environment variable.

You can run tests after [building a specific stage](bootstrap.md) by

> adding the `-C stageN` argument. Most of the tests are in stage1.  The

Reviewer: ```suggestion
adding the `-C stageN` argument. The default when run as above is stage 1.  The
```

Lean tests will automatically use that stage's corresponding Lean
executables

You can also use `ctest` directly if you are in the right folder.  So
to run stage1 tests with a 300 second timeout run this:

```bash
cd build/release/stage1
ctest -j 4 --output-on-failure --timeout 300
```

> To get verbose output from ctest pass the `-VV` command line option.

Reviewer: Interesting, I don't think I've ever used that. When should we use it?

Reviewer: In a quick test, the output from `-V` and `-VV` looks identical to me

Author: The [documentation ](https://cmake.org/cmake/help/v3.0/manual/ctest.1.html) says 

- Enable more verbose output from tests.
   Test output is normally suppressed and only summary information is displayed. This option will show _even more_ test output.

But it could be that our style of testing there is no "even more" output available... so I updated the doc to just mention `--verbose` and what that does exactly.

Author: for the record I did find some differences when comparing the entire output with -V versus -VV :-) 

![image](https://user-images.githubusercontent.com/18707114/134433688-b64633d0-b571-4cf2-b857-d25cc19e021b.png)

Reviewer: I think that's just `make` skipping work on a repeated run

Note that is two letter `V`'s and not a `W`.  I guess it means very
verbose.

Here is the summary of the test source code organization.

> All these tests are included by [/src/shell/CMakeLists.txt](https://github.com/leanprover/lean4/blob/master/src/shell/CMakeLists.txt):

Reviewer: I don't think we use this leading slash convention anywhere else

Author: fixed
